### ISSUE TITLE
Recurring cards

### ISSUE THREAD

**Issue by Cante** [nextcloud/deck#480]
It should be possible to define a recurrence interval.
For example: weekly repeat on monday at 20:00

**Comment by Cante** [comment]
This is true. You can close the ticket.

**Comment by jakobroehrl** [comment]
Other Apps solve this so:
When a recurrent task/card is archived/set to done it will pop up again in the future with a new due date

**Comment by juliushaertl** [comment]
Might be something once we have #534

**Comment by horvan** [comment]
Would like to se this Feature, soon. Especially since dashboard shows cards they are a cool alternative  to a full calendar. So should it be hard to archive cards and reopen them again if they have a recurrent option?

Plus 1 for this. I like the Idea.

**Comment by l-mb** [comment]
Indeed, I'd love the ability to mark a card as done, and for it to then automatically reappear with the new due date as per the recurring interval. Possibly with the option of being put to the end of the list, or even to another list.
So there'd always just be "one" card in the deck, not multiple.

Right now, we can't even have recurring tasks, and so have to manage that manually or externally.

This would greatly help with workflow management.

Thanks for the consideration!

**Comment by eX00r** [comment]
Any news on this?

**Comment by DDriggs00** [comment]
This would be a great feature.

**Comment by DriedPlumber** [comment]
I just noticed that I would like too to have a feature for "recurring" Deck cards...
I have just created a card to pay bills and I realize that I will have to do it every month...
Thx

**Comment by daniel-sousa-me** [comment]
> Other Apps solve

> Right now, we can't even have recurring tasks, and so have to manage that manually or externally.

What other apps or tools can help with this?

**Comment by Extarys** [comment]
@meichthys I currently use a specific calendar for recurrence, and add emoji at the start of the event's name to see what's what at a glance, but it's a mess. I have 10+ events per day sometimes. (add the regular calendar events and you get eyestrain)

**Comment by viktorix** [comment]
Just wanted to add a bit more information as I would love to see this feature. For personal stuff, I've used Nifty project management app. Here's their UI to create a recurring card:
![image](https://github.com/nextcloud/deck/assets/1692600/a07e78a3-190c-4469-a42d-d26775c9b94e)

When a card is completed, a new task is created in the column where original recurring card was created. So if I create a recurring card in column "To Do", and then it is completed in the column "Done", the new card will be created in "To Do" column.

**Comment by helheimcloud** [comment]
@viktorix  Additionally, it would be nice if you could have a setting  "Show card X days before due date". It would not make sense to have the task pop up again just after closing it. This would then be a nice minimalistic "Preventive Maintaince Task Scheduling" System. 

I hope this will be added soon. I really want to use Deck instead of an external tool. People are requesting it since 2018 anyway :D

**Comment by marcoambrosini** [comment]
Related https://github.com/nextcloud/tasks/pull/2318

**Comment by hansimausisteffiwurstbrot** [comment]
> Related [nextcloud/tasks#2318](https://github.com/nextcloud/tasks/pull/2318)

Kind of, but about another app.. :)

**Comment by DinoBellini** [comment]
> It should be possible to define a recurrence interval. For example: weekly repeat on monday at 20:00

example: every 6 month first monday; every week on friday, first and third monday of month.....

**Comment by distresscode** [comment]
any progress on this?

**Comment by juliushaertl** [comment]
@nimishavijay Are you already working on something there for design specs or does it make sense to have some kind of kick-off meeting? So far we don't have any clear requirement definition.

**Comment by juliushaertl** [comment]
For inspiration https://support.atlassian.com/trello/docs/using-the-card-repeater-power-up/

Two ways of handling recurrence:
- Use the same card and just set the due date again (like caldav tasks do it)
- Clone the card to a new one with the recurrence (how trello does it)

We should get an overview on how others handle it, what the most common path is

**Comment by nimishavijay** [comment]
So taking a look at Trello and [Asana](https://github.com/user-attachments/assets/c7cc9d82-5ba6-4221-84b8-dffa6b0b6074), the card is cloned to a new one. Which makes sense in most use cases of this feature (Eg. a recurring card for a weekly meeting where the description is the agenda, and cloning the card would help to prevent loss of the previous week's agendas and action items). 

The main difference between Asana and Trello is that in Trello the repeating and due date are separate, setting a card as repeating does not affect its due date, the card is literally cloned. In Asana, the setting a repeat means that the card is cloned with a new due date for every card. 

Like many of the above comments it would be great to set the due date as a part of the recurrence, it could be cumbersome to have a card pop up automatically but you have to calculate and set the due date, especially when you're out of the original context. 

## UI
The calendar is a good reference for the UI itself.
- [ ] Below the row with the due date, there could be a "🔁 does not repeat" and a "+" button, clicking on which opens the recurrence selector.
- [ ] The recurrence selector design could be similar to the (intended) design in Calendar https://github.com/nextcloud/calendar/issues/4618#issuecomment-1914583274 with the "End repeat" as a nice-to-have
  - [ ] One important change is a little subline in the dialog with a short explanation about the recurrence. "When a card is marked as done, another one will be created with the set due date" or something similar
  - [ ] There should also be an dropdown to set the column that it repeats in. The default value is the column the card is in currently, but if the card is moved to a different board and modified then it should be clear which column the card will repeat in. 
- [ ] When a card is recurring, the chip with the due date can have a 🔁 symbol next to the clock symbol
- [ ] Icon: Material Symbols [repeat icon](https://fonts.google.com/icons?selected=Material+Symbols+Outlined:repeat:FILL@0;wght@400;GRAD@0;opsz@24&icon.query=repeat&icon.size=24&icon.color=%23222222)


## Behaviour 
- [ ] if recurrence is set on a card that already has a due date, the recurrence starts after the due date (eg. due date is 15.10 (Tuesday) and recurrence is weekly on Monday => first due date is 15.10 (tuesday), then 21.10 (Monday), 28.10 (Monday), etc)
- [ ] once recurrence is set for card A,
  - [ ] due date is changed => change only due date od 
  - [ ] card A is marked as done => card B with new due date occurs in the original column where the recurrence was set
  - [ ] card A is deleted => modal with message "Deleting a card also stops it from repeating. To keep repeating the card, archive it instead" with options to archive, delete, and cancel

Let me know what you think!

**Comment by craftypenguins-chris** [comment]
This behaviour has been implemented in other platforms in a less-than-ideal way.

In MS Planner, for example, marking a recurring task as done with creating a new task becomes quite cumbersome. This is especially true with large teams pulling the tasks from the on deck in true KanBan style.

I propose an alternate logic for this, which Nextcloud codebase may or may not be suitable for. If not, this may need to be a separate app altogether:

Recurring tasks would be treated like a set, grouped so that the UX for managing them has the functions needed to control all tasks in the set.

What would be great would be to have a generation trigger option which specifies when the user would like the next task created. For example, if we have a task that is monthly. That task could be set to regenerate on the 1st day of the month. A weekly task can specify which day of the week to trigger the next set. Then, a checkbox to decide generation based on the previous task being completed (for example).

Another advantage to managing the recurring tasks as a set is that you can manipulate the group, i.e., pause the set or delete all tasks up to a certain date.

Happy to invest some time into this as we are considering this product for our internal business use case and moving away from M365. Also, open to alternative thoughts on how to accomplish the above. As a user of Planner and other task tools, I can tell you this is a bother for most everyone.

**Comment by mhzawadi** [comment]
Hello 👋🏼 

Could this be a flow that you can enable?

**Comment by Cyberseal89** [comment]
Would like to see it when it completes and archives for an automated card be added based on the periodicity due date to be set. I use deck for maintenance, and it’s extremely helpful, I complete than have to assign a new due date like 3 month out = quarter, with a new due date.

**Comment by Mrs-Feathers** [comment]
still waiting for reoccuring tasks

**Comment by janvv** [comment]
Would love to see it too. I use Trello Card Repeater Powerup. Is there something similar like Powerups for Cards?

**Comment by silbert3** [comment]
Hi, I just discovered Nextcloud Deck and I love it. Only thing I am missing is to have recurring tasks, favorably the possibility to auto- create a new card from the last one with a due date in X amount of days after closure. Happy do find this discussion - thanks a lot, guys!

**Comment by mhzawadi** [comment]
could this be part of the done flag stuff?

**Comment by fernandorogerio** [comment]
The **Tasks** app already displays **Deck** cards but does not yet allow setting up recurrence for them, even though the feature has been available for native Tasks items since the latest release. It might be possible to enable this functionality by adjusting the Deck card properties to support recurrence.

https://github.com/nextcloud/tasks/pull/3021

**Comment by aktasfatih** [comment]
I made a quick https://github.com/aktasfatih/deck_recurrence app if anyone is interested in. 

<img width="895" height="760" alt="Image" src="https://github.com/user-attachments/assets/e8eebc67-2988-48c4-8e1e-06d6e73a9966" />

**Comment by jcharaoui** [comment]
> I made a quick https://github.com/aktasfatih/deck_recurrence app if anyone is interested in.

Thanks, this looks interesting. I was tempted to try it, until I understood that "I made a quick" was actually code for "I used AI for".

**Comment by aktasfatih** [comment]
I can understand where you are coming from 👍 If anyone finds any problems in my repo feel free to create an issue or make a PR. I will probably build many of the other features I am looking for this way. It'd be good to get it to an acceptable standard. Thanks 👍

**Comment by aktasfatih** [comment]
I have expanded its features and added more options. You can keep using the same card, create when the previous one is done and also uncheck the checklist now. Also, added a `Create now` button and an OCS API for it.

<img width="1006" height="950" alt="Image" src="https://github.com/user-attachments/assets/3168a484-c1c8-4a12-a6be-e1c74e472444" />

**Comment by dbmtrde** [comment]
> I made a quick https://github.com/aktasfatih/deck_recurrence app if anyone is interested in.
> 
> [<img alt="Image" width="895" height="760" src="https://private-user-images.githubusercontent.com/12985956/622243853-e8eebc67-2988-48c4-8e1e-06d6e73a9966.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODU4Mzk0MTQsIm5iZiI6MTc4NTgzOTExNCwicGF0aCI6Ii8xMjk4NTk1Ni82MjIyNDM4NTMtZThlZWJjNjctMjk4OC00OGM0LThlMWUtMDZkNmU3M2E5OTY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA4MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwODA0VDEwMjUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhOGYzZjQ4NDg3MzU3ZmVlYjM0YjBmZTJmYWNhODkyYjIxNTU5MmJmYzQxYmJmMjJkMmJiZTU5MDk3MjNkMDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.q3nG1rg7X4Hde0N9DEM9GINa4e4mAZwSsYfQqrjppWM">](https://private-user-images.githubusercontent.com/12985956/622243853-e8eebc67-2988-48c4-8e1e-06d6e73a9966.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODU4Mzg3MzQsIm5iZiI6MTc4NTgzODQzNCwicGF0aCI6Ii8xMjk4NTk1Ni82MjIyNDM4NTMtZThlZWJjNjctMjk4OC00OGM0LThlMWUtMDZkNmU3M2E5OTY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA4MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwODA0VDEwMTM1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQyMDlkZjNiNTgzY2RkMzFhYTY1NmEwYWM3YTdmNmE1ZmI5MGE5ODdiZDY1ZjYzM2ExZmU4ZmUyNjYzNDE5YjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.X3HA86Yu3lKA8T2UYn5mM9pZA8uiIswNbaMCPRdIEWM)

Thanks for the feature/app @aktasfatih  it looks promising. 🫶 
Before I start using it in a production environment: Are there plans to make it available in the Nextcloud App Store? If so, would switching from a manual `custom_apps` installation to the Store version later on work without any data loss (rules, cards that have already been generated, etc.), or would that require a fresh installation?

**Comment by aktasfatih** [comment]
@dbmtrde Thanks 👍 No, I currently have no plans in putting `deck_recurrence` in the store. Also, you still wouldn't lose data even if I put it in the store. You should always make sure you take backups though just in case something goes wrong🤞 

Nevertheless, after building `deck_recurrance` I realized building separate apps to have the missing features I am looking for was such an anti-pattern. I also kept wanting features that didn't exist and the list would just grow. Some features were also just not doable through other apps. So, instead now I built a whole new app now. Feel free to check it out https://github.com/aktasfatih/kanso. It has a ton of the stuff people have asked for. (recurring cards, card templates, WIP limits, public board sharing, a calendar feed, automation rules, a timeline/Gantt view, analytics, and more).

It has board/deck importers so you can switch over easily as well. 

It is currently waiting on https://github.com/nextcloud/app-certificate-requests/pull/1129 to go on the store. Happy to take feedback, bug reports or new ideas on it 👍

**Comment by mhzawadi** [comment]
@aktasfatih that looks amazing, but nextcloud 30-32 would put you 2 major versions behind. That might not pass the app store checks.

I would be interested if it worked on nextcloud 34 and PHP 8.3

**Comment by aktasfatih** [comment]
I will look into it right now. Thanks for raising it 👍 

Edit: done

**Comment by dbmtrde** [comment]
> [@dbmtrde](https://github.com/dbmtrde) Thanks 👍 No, I currently have no plans in putting `deck_recurrence` in the store. Also, you still wouldn't lose data even if I put it in the store. You should always make sure you take backups though just in case something goes wrong🤞
> 
> Nevertheless, after building `deck_recurrance` I realized building separate apps to have the missing features I am looking for was such an anti-pattern. I also kept wanting features that didn't exist and the list would just grow. Some features were also just not doable through other apps. So, instead now I built a whole new app now. Feel free to check it out https://github.com/aktasfatih/kanso. It has a ton of the stuff people have asked for. (recurring cards, card templates, WIP limits, public board sharing, a calendar feed, automation rules, a timeline/Gantt view, analytics, and more).
> 
> It has board/deck importers so you can switch over easily as well.
> 
> It is currently waiting on [nextcloud/app-certificate-requests#1129](https://github.com/nextcloud/app-certificate-requests/pull/1129) to go on the store. Happy to take feedback, bug reports or new ideas on it 👍

Your Kanso app also looks promising and, at first glance, would address many of the Deck's shortcomings, so it would be great if you could release it as an app in the Store. I'd definitely be one of the first to use it as soon as it's available in the Store. 👍

**Comment by mhzawadi** [comment]
Can confirm Kanso is amazing, @aktasfatih has done an amazing job getting that working.

I would recommend installing and having a play if your at all interested

**Comment by aktasfatih** [comment]
Hi, the app is on the nextcloud app store. https://apps.nextcloud.com/apps/kanso I hope it helps anyone looking for this feature in NextCloud! Thanks 👍

**Comment by dbmtrde** [comment]
> Hi, the app is on the nextcloud app store. https://apps.nextcloud.com/apps/kanso I hope it helps anyone looking for this feature in NextCloud! Thanks 👍

What a great job! As promised, I’m one of the first users, and it looks really good. 😉👌
Thanks a lot!

PS: I’d also love to see deck_recurrence in the App Store as well.
